File: C/Linked-List/Reverse a singly linked list.c
Reversing a linked list using iterative approach

#include<stdio.h>

#include<stdlib.h>

//Node structure of the linked list

struct Node{
    
  int value;
    
  struct Node* next;

};


//Reverse the linked list function

void reverse(struct Node** head)

{
    
  struct Node* p=NULL;
    
  struct Node* c=*head;
    
  struct Node* next=NULL;
    
  while(c!=NULL){
     
   //stroring the next address of current node
    next=c->next;
     
    //reversing the current node's ponter
    c->next=p;
    
    //moving the pointer on position ahead
    p=c;
        
    c=next;
    
  }
    
  *head=p;

}

//To push a node in linked list

void push(struct Node** head, int val){

  struct Node* newnode=(struct Node*)malloc(sizeof(struct Node));

  newnode->value=val;

  newnode->next=(*head);

  *head=newnode;

}

//Printing function-to print the linked list

void show(struct Node* h){
    
  struct Node* t=h;
    
  while(t!=NULL){
        
    printf("%d ", t->value);
        
    t=t->next;
    
  }


}



int main()

{
//empty list
  struct Node* h = NULL;


  push(&h, 25);

  push(&h, 36);

  push(&h, 44);

  push(&h, 50);

  push(&h, 17);

  push(&h, 66);


  printf("Linked list before reversal:\n");

  show(h);

  reverse(&h);

  printf("\nLinked list after reversal:\n");

  show(h);

}

Time Complexity of reversing a linked list is O(n) 

File: Python/Linked-List/Reverse a singly linked list.py
Reversing a linked list using iterative approach
#node of singly linked list

class Node:
    
    #constructor used to initialize the node object
    
    def __init__(self, value):
        
        self.value = value
        
        self.next = None


class l_list:

    #head initialization
    
    def __init__(self):
        
        self.head = None
        
        self.l_node = None

# inserting a new node at the beginning
    
  def append(self,value):
        
      if self.l_node is None:
            
          self.head = Node(value)
            
          self.l_node = self.head
        
      else:
            
          self.l_node.next = Node(value)
            
          self.l_node = self.l_node.next

      #function to print the linkedlist
    
    def show(self):
        
        c=self.head
        
        while c:
            
            print(c.value, end=' ')
            
            c=c.next

#function to reverse the linked list

def reverse(linked_list):
    
    prev = None
    
    c = linked_list.head
    
    if c is None:
        
        return
    
    after = c.next
    
    while after:
        
        c.next = prev
        
        prev = c
        
        c = after
        
        after = after.next
    
    c.next = prev
    
    linked_list.head = c

#driver code

linked_list = l_list()

a=int(input("Enter number of elements of linked list: "))

print("Enter data of linked list one at a time: ")

for i in range(a):
    
    value=int(input())
    
    linked_list.append(value)


reverse(linked_list)

print("The reversed linked list is: ")

linked_list.show()

Time Complexity of reversing a linked list is O(n)  
